File: COSC 4315/assign1/JavaScript/is-Perfect.js
function isPerfectImperative(number){
    let sum = 0;
    for (let i = 0; i <= number; i++){
        if (number % i === 0){
            sum += i;
        }
    }
    return sum === 2 * number;
}

function isPerfectFunctional(number){
    return number * 2 ===
        new Array(number)
            .fill()
            .map((index, i) => i + 1)
            .filter(i => number % i === 0)
            .reduce((sum, i) => sum + i, 0);
}

let numbers = [1, 5, 6, 7, 28, 496];

console.log("Imperative Style:");

for (const i of numbers) {
    let result = isPerfectImperative(i) ? "Yes" : "No";
    console.log(`${i} : ${result}`);
}

console.log("\nFunctional Style:");

for (const i of numbers) {
    let result = isPerfectFunctional(i) ? "Yes" : "No";
    console.log(`${i} : ${result}`);
}

File: COSC 4315/assign1/swift/isPerfect.swift
func isPerfectImperative(number: Int) -> Bool {
    var sum = 0
    for index in 1...number {
        if (number % index == 0) {
            sum += index
        }
    }
    return sum == 2 * number
}

func isPerfectFunctional(number: Int) -> Bool {
    return number == (1..<number)
      .filter({number % $0 == 0})
      .reduce(0, {$0 + $1})
}

let numArray = [1, 5, 6, 7, 28, 496]

print("Imperative style:")

for index in numArray {
    let result = isPerfectImperative(number: index) ? "Yes" : "No"
    print("\(index): \(result)")
}

print("")

print("Functional style:")

for index in numArray {
    let result = isPerfectFunctional(number: index) ? "Yes" : "No"
    print("\(index): \(result)")
}
